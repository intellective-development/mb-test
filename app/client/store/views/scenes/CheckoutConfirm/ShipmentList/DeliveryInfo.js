// @flow

import * as React from 'react';
import { delivery_method_helpers } from 'store/business/delivery_method';

import { MBText } from '../../../elements';

type DeliveryInfoProps = {shipment: Object};
class DeliveryInfo extends React.PureComponent<DeliveryInfoProps> {
  render(){
    const { delivery_method } = this.props.shipment;

    return (
      <div className="csl__delivery__container">
        <div className="csl__delivery__name-container">
          <MBText.Span className="csl__delivery__name">
            {delivery_method_helpers.displayName(delivery_method)}
            <ClosedWarning delivery_method={this.props.shipment.delivery_method} />
          </MBText.Span>
        </div>
        <MBText.Span className="csl__delivery__next-delivery">
          {delivery_method_helpers.formatNextDelivery(delivery_method)}
        </MBText.Span>
      </div>
    );
  }
}

const ClosedWarning = ({delivery_method}) => {
  if (!delivery_method_helpers.isClosed(delivery_method)) return null;

  return <MBText.Span className="csl__delivery__closed">&ensp;Closed</MBText.Span>;
};

export default DeliveryInfo;
