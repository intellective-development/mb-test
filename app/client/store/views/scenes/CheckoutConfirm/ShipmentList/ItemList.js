// @flow

import * as React from 'react';
import { product_grouping_helpers } from 'store/business/product_grouping';
import formatCurrency from 'shared/utils/format_currency';
import { cart_item_helpers } from 'store/business/cart_item';

import { MBText, MBTouchable } from '../../../elements';

const LIST_INITIAL_LENGTH = 3;

type ItemListProps = { items: Array<Object> };
type ItemListState = { show_all: boolean };
class ItemList extends React.PureComponent<ItemListProps, ItemListState> {
  state = {show_all: false}

  showAllItems = () => {
    this.setState({show_all: true});
  }

  render(){
    const { items } = this.props;
    const { show_all } = this.state;
    const has_hidden_items = items.length > LIST_INITIAL_LENGTH && !show_all;

    let display_items = items;
    if (!show_all) display_items = items.slice(0, LIST_INITIAL_LENGTH);

    return (
      <div className="csl__item-list-container">
        {display_items.map(item => (
          <ItemListElement item={item} key={item.variant.id} />
        ))}
        <ShowAll is_hidden={!has_hidden_items} showAllItems={this.showAllItems} />
      </div>
    );
  }
}

const ItemListElement = ({item}) => {
  const {variant, product_grouping, quantity} = item;
  let volume_el = '';
  if (variant.volume){
    volume_el = <span>&nbsp;&ndash;&nbsp;{variant.volume}</span>;
  }

  return (
    <div className="csl__item-container">
      <img
        alt={product_grouping.name}
        src={product_grouping_helpers.getThumb(product_grouping, variant)}
        className="csl__item-image" />
      <div className="csl__item-detail__container">
        <MBText.Span className="csl__item-detail__text">
          {product_grouping.name}
        </MBText.Span>
        <MBText.Span className="csl__item-detail__text">
          Qty: {quantity}&nbsp;&ndash;&nbsp;{formatCurrency(cart_item_helpers.itemSubtotal(item))}{volume_el}
        </MBText.Span>
      </div>
    </div>
  );
};

const ShowAll = ({is_hidden, showAllItems}) => {
  if (is_hidden) return null;

  return (
    <MBTouchable onClick={showAllItems} className="secondary-action show-all">Show all items</MBTouchable>
  );
};

export default ItemList;
