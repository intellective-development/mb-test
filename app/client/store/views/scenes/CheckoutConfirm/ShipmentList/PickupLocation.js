// @flow

import * as React from 'react';
import { connect } from 'react-redux';
import I18n from 'store/localization';
import { address_helpers } from 'store/business/address';
import { delivery_method_constants } from 'store/business/delivery_method';
import { supplier_helpers } from 'store/business/supplier';
import type { Shipment } from 'legacy_store/models/Shipment';
import { ui_actions } from 'store/business/ui';

import { MBText, MBTouchable } from '../../../elements';

type PickupLocationProps = {shipment: Shipment, showSupplierMap: Function};
export const PickupLocation = ({shipment, showSupplierMap}: PickupLocationProps) => {
  if (shipment.delivery_method.type !== delivery_method_constants.PICKUP) return null;

  return (
    <MBTouchable className="csl__pickup__container" onClick={() => showSupplierMap(shipment.supplier.id)} >
      <MBText.P className="csl__pickup__section-label">{I18n.t('ui.body.checkout_shipment.pickup_location_prompt')}</MBText.P>
      <MBText.P className="csl__pickup__address">
        {address_helpers.formatStreetAndCity(shipment.supplier.address)}
        ãƒ»{supplier_helpers.formatDistance(shipment.supplier)}
      </MBText.P>
    </MBTouchable>
  );
};

const PickupLocationDTP = {showSupplierMap: ui_actions.showSupplierMapModal};
const PickupLocationContainer = connect(null, PickupLocationDTP)(PickupLocation);

export default PickupLocationContainer;
