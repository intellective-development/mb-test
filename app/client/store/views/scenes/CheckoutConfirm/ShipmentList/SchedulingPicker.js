// @flow

import * as React from 'react';
import _ from 'lodash';
import moment from 'moment';
import { connect } from 'react-redux';
import cn from 'classnames';
import I18n from 'store/localization';
import * as Ent from '@minibar/store-business/src/utils/ent';
import * as date_helpers from '@minibar/store-business/src/utils/format_date';
import { delivery_method_helpers } from 'store/business/delivery_method';
import { order_actions } from 'store/business/order';
import * as shipment_helpers from 'legacy_store/models/Shipment';
import { scheduling_calendar_actions } from 'store/business/scheduling_calendar';
import type { SchedulingCalendar } from 'store/business/scheduling_calendar';

import { MBText } from '../../../elements';

type SchedulingPickerProps = {
  shipment: Object,
  scheduling_calendar: SchedulingCalendar,
  resetShipmentScheduling: Function,
  setShipmentScheduling: Function,
  fetchCalendar: Function
};
type SchedulingPickerState = {loading_calendar: boolean, selected_date: ?string};
class SchedulingPicker extends React.PureComponent<SchedulingPickerProps, SchedulingPickerState> {
  constructor(props: SchedulingPickerProps){
    super(props);

    if (!props.scheduling_calendar){
      props.fetchCalendar(props.shipment.delivery_method.id);
      this.state = {loading_calendar: true, selected_date: null};
    } else {
      const initial_date = this.initialSchedulingDay(props.shipment.scheduled_for, props.scheduling_calendar);
      this.state = {loading_calendar: false, selected_date: initial_date};
    }
  }

  componentWillReceiveProps(next_props){
    if (!this.props.scheduling_calendar && next_props.scheduling_calendar){
      const initial_date = this.initialSchedulingDay(next_props.shipment.scheduled_for, next_props.scheduling_calendar);
      this.setState({loading_calendar: false, selected_date: initial_date});
    } else if (this.props.scheduling_calendar && !next_props.scheduling_calendar){
      next_props.fetchCalendar(next_props.shipment.delivery_method.id);
      this.setState({loading_calendar: true, selected_date: null});
    }
  }

  initialSchedulingDay = (scheduled_for, scheduling_calendar) => {
    if (!scheduled_for) return null;

    const scheduled_for_moment = moment.parseZone(scheduled_for);
    const scheduled_day = scheduling_calendar.scheduling_days.find(day => (
      moment.parseZone(day.windows[0].start_time).isSame(scheduled_for_moment, 'day')
    ));
    return scheduled_day ? scheduled_day.date : null;
  }

  selectSchedulingDay = (date: string) => {
    this.props.resetShipmentScheduling(!!date, this.props.shipment.supplier.id);
    this.setState({selected_date: date});
  }

  selectSchedulingWindow = (scheduled_for: string) => {
    this.props.setShipmentScheduling(scheduled_for, this.props.shipment.supplier.id);
  }

  render(){
    const { selected_date, loading_calendar } = this.state;
    const { shipment, scheduling_calendar } = this.props;
    const { delivery_method } = shipment;

    if (!delivery_method_helpers.deliveryMethodCanBeScheduled(delivery_method)) return null;
    if (loading_calendar) return <SchedulingLoader />;

    const selected_scheduling_day = scheduling_calendar.scheduling_days.find(day => day.date === selected_date);
    const must_be_scheduled = delivery_method_helpers.mustBeScheduled(delivery_method);
    const scheduling_windows = selected_scheduling_day ? selected_scheduling_day.windows : [];

    return (
      <form className="csl__scheduling__container">
        <SchedulingLabel shipment={shipment} />
        <div className="csl__scheduling__select-container">
          <DateSelect
            scheduling_dates={scheduling_calendar.scheduling_days.map(day => day.date)}
            selected_date={selected_date}
            required={must_be_scheduled}
            selectSchedulingDay={this.selectSchedulingDay} />
          <TimeSelect
            windows={scheduling_windows}
            selected_window={shipment.scheduled_for}
            disabled={!selected_scheduling_day}
            selectSchedulingWindow={this.selectSchedulingWindow} />
        </div>
      </form>
    );
  }
}

const SchedulingLabel = ({shipment}) => {
  const error = shipment_helpers.validateShipment(shipment);
  const content_path = error ? error.name : 'scheduling_prompt';
  const classes = cn('csl__scheduling__section-label', {
    'csl__scheduling__section-label--error': !!error
  });

  return (
    <MBText.P className={classes}>
      {I18n.t(`ui.body.checkout_shipment.${content_path}`, {
        delivery_method_type: delivery_method_helpers.displayNameShort(shipment.delivery_method).toLowerCase()
      })}
    </MBText.P>
  );
};

const DateSelect = ({scheduling_dates = [], selected_date, required, selectSchedulingDay}) => {
  const default_prompt = required ? 'Select Date' : 'ASAP';
  const default_value = '';

  return (
    <select
      name="date"
      className="csl__scheduling__date-select"
      value={selected_date || default_value}
      onChange={e => selectSchedulingDay(e.target.value)}>
      <option value={default_value}>{default_prompt}</option>
      {scheduling_dates.map(date => (
        <option value={date} key={date}>{formatWindowDate(date)}</option>
      ))}
    </select>
  );
};

const TimeSelect = ({windows = [], selected_window, disabled, selectSchedulingWindow}) => {
  const default_value = '';

  return (
    <select
      name="time"
      className="csl__scheduling__time-select"
      disabled={disabled}
      value={selected_window || default_value}
      onChange={e => selectSchedulingWindow(e.target.value)}>
      <option value={default_value} disabled style={{display: 'none'}}>Select Time</option>
      {windows.map(window => (
        <option value={window.start_time} key={window.start_time}>
          {date_helpers.formatTime(window.start_time)} - {date_helpers.formatTime(window.end_time)}
        </option>
      ))}
    </select>
  );
};

const SchedulingLoader = () => {
  return <div className="csl__scheduling__loader" />;
};

const SchedulingPickerSTP = () => {
  const findCalendar = Ent.find('scheduling_calendar');

  return (state, {shipment}) => ({
    scheduling_calendar: findCalendar(state, shipment.delivery_method.id)
  });
};
const SchedulingPickerDTP = {
  resetShipmentScheduling: order_actions.resetShipmentScheduling,
  setShipmentScheduling: order_actions.setShipmentScheduling,
  fetchCalendar: scheduling_calendar_actions.fetchCalendar
};
const SchedulingPickerContainer = connect(SchedulingPickerSTP, SchedulingPickerDTP)(SchedulingPicker);

export default SchedulingPickerContainer;

// helpers

// returns whether ISO 8601 timestamp is today in it's own timezone
export const isToday = (date: string) => {
  const date_moment = moment.parseZone(date);
  return moment(date_moment).format('YYMMDD') === moment().format('YYMMDD');
};

const formatWindowDate = (date) => {
  if (!date) return '';
  const month_and_day = moment.parseZone(date).format('MMM D');
  const day_name = isToday(date) ? _.startCase(I18n.t('global.day.today')) : _.startCase(date_helpers.formatDayName(date, {short: true}));
  return `${day_name}, ${month_and_day}`; // e.g. Today, Aug 22 || Thu, Aug 24
};
