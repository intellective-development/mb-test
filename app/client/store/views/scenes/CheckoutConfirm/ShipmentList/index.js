// @flow

import * as React from 'react';
import I18n from 'store/localization';
import * as Ent from '@minibar/store-business/src/utils/ent';
import { connect } from 'react-redux';
import connectToObservables from 'shared/components/higher_order/connect_observables';
import { orderStream } from 'legacy_store/models/Order';
import { order_helpers } from 'store/business/order';
import { cart_item_selectors } from 'store/business/cart_item';
import { supplier_selectors } from 'store/business/supplier';
import * as shipment_helpers from 'legacy_store/models/Shipment';
import formatCurrency from 'shared/utils/format_currency';
import { analytics_actions /*, analytics_helpers */} from 'store/business/analytics';

import { MBText } from '../../../elements';
import DeliveryInfo from './DeliveryInfo';
import ItemList from './ItemList';
import PickupLocation from './PickupLocation';
import SchedulingPicker from './SchedulingPicker';

export class ShipmentList extends React.PureComponent {
  /*componentDidMount(){
    const { cart_items = [], trackEvent } = this.props;
    TODO: cart_items.forEach(({ product_grouping, variant, quantity }) => {
      trackEvent({
        action: 'checkout_place_order_in_cart',
        content_type: 'product',
        items: [analytics_helpers.getCartItemData(product_grouping, variant, quantity)]
      });
    });
  }*/

  render(){
    const {shipments = []} = this.props;
    return (
      <div>
        {shipments.map(shipment => (
          <ShipmentListItem
            shipment={shipment}
            key={shipment.supplier.id} />
        ))}
      </div>
    );
  }
}

type ShipmentListItemProps = {shipment: Object};
const ShipmentListItem = ({shipment}: ShipmentListItemProps) => {
  return (
    <div className="panel-group">
      <div className="panel heading">
        <ShipmentHeader shipment={shipment} />
      </div>
      <div className="panel">
        <DeliveryInfo shipment={shipment} />
        <PickupLocation shipment={shipment} />
        <SchedulingPicker shipment={shipment} />
      </div>
      <div className="panel">
        <ItemList items={shipment.items} />
      </div>
    </div>
  );
};

const ShipmentHeader = ({shipment}) => {
  return (
    <div>
      <MBText.H2 className="csl__panel_heading">{shipment.supplier.name}</MBText.H2>
      <MBText.H3 className="csl__panel_subheading">
        {I18n.t('ui.body.checkout_shipment.items', {
          count: shipment_helpers.shipmentItemCount(shipment) || 0
        })}
        &ensp;&ndash;&ensp;
        {formatCurrency(shipment_helpers.shipmentSubtotal(shipment))}
      </MBText.H3>
    </div>
  );
};

export const ShipmentListSTP = () => {
  const findCartItems = Ent.query(Ent.find('cart_item'), Ent.join('variant'), Ent.join('product_grouping'));
  const findSuppliers = Ent.query(Ent.find('supplier'), Ent.join('delivery_methods'));
  return (state, {order}) => {
    if (!order){
      return {
        shipments: [],
        cart_items: cart_item_selectors.getAllCartItems(state)
      };
    }
    const cart_items = findCartItems(state, cart_item_selectors.getAllCartItemIds(state));
    const suppliers = findSuppliers(state, supplier_selectors.currentSupplierIds(state));
    const selected_delivery_methods = supplier_selectors.selectedDeliveryMethods(state);
    return {
      shipments: order_helpers.getOrderShipments(order, cart_items, suppliers, selected_delivery_methods),
      cart_items: cart_item_selectors.getAllCartItems(state)
    };
  };
};

const ShipmentListDTP = {
  trackEvent: analytics_actions.track
};

export const ShipmentListContainer = connect(ShipmentListSTP, ShipmentListDTP)(ShipmentList);

export default connectToObservables(ShipmentListContainer, {order: orderStream});
