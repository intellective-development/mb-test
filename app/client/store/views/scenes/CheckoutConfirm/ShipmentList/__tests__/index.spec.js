import * as React from 'react';
import TestProvider from 'store/views/__tests__/utils/TestProvider';

import ShipmentList from '../index';

jest.mock('legacy_store/collections/Suppliers');
jest.mock('legacy_store/models/Cart');
jest.mock('legacy_store/models/Order');
jest.mock('client/shared/components/higher_order/make_provided'); // avoid importing the full store

describe('ShipmentList', () => {
  // TODO: test with actual shipment state

  it('renders', () => {
    expect(render(
      <TestProvider>
        <ShipmentList />
      </TestProvider>
    )).toMatchSnapshot();
  });
});
