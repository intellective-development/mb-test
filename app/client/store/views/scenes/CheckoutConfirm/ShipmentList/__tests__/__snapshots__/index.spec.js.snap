// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShipmentList renders 1`] = `<div />`;
